### ISSUE TITLE
Add support for multi-line lints

### ISSUE THREAD

**Issue by artificerchris** [semagnum/blint#34]
The title pretty much says it all, I'd like to add lints that need to span multiple lines because they need to do things like `import os` which can't be done on a single line of python.

**Comment by semagnum** [comment]
With complex issues, the best workaround currently is writing a full script (either as its own file or a .blend file's text block) and running within the file. You could have blender.exe run that script directly in the command line (see [Blender docs](https://docs.blender.org/manual/en/latest/advanced/command_line/arguments.html#python-options) for that).

For issue expressions, you are correct that multiline is unsupported. BLint uses `eval()`, so it only supports a single expression. That's partially for security reasons, and mitigates the likelihood of side effects just by evaluating. I got around the import problem by using `__import__` with the built-in config's missing texture issue:

```json
"issue_expr": "not __import__('os').path.isfile(bpy.path.abspath(my_image.filepath)",
```

But I have a feeling you're not using `os` once 😉

If you are referring to issue fixes, multiline is supported. BLint uses `exec()`. You can use newlines, tabs, or semicolons to split statements up. Like in this `fix_expr`:

```json
        {
            "description": "Multiline test",
            "severity_icon": "INFO",
            "category_icon": "IMAGE_DATA",
            "issue_expr": "my_image.source not in {'GENERATED', 'VIEWER'}",
            "prop_label_expr": "my_image.name",
            "iterable_expr": "bpy.data.images",
            "iterable_var": "my_image",
            "fix_expr": "import os; print(os.path.isfile(my_image.filepath))"
        }
```

BLint is already doing multiple statements with iterative fixes:
```bash
INFO:bl_ext.blender_scripts.blint.operators.issue_fix:Running: my_image = [my_image for my_image in bpy.data.images if my_image.source not in {'GENERATED', 'VIEWER'}][0]; import os; print(os.path.isfile(my_image.filepath))
True
```

Of course, code all on one line isn't always pretty 🙂 Even if I added multiline support for issue detection, there's two other problems to take into account:

- the storage format. JSON doesn't support multiline strings. So even if we supported multiline with issue detection, you'd still have to add separators in the JSON. And no one would want to write that, coder or not. To support multiline directly in the storage means changing the file format.
- Blender API. The `StringProperty` may save newlines, but it's stripped when displayed in the UI. That means newlines cannot be added in the BLint editor in the preferences. So editing the code to have newlines would _have_ to take place either in the Scripting editor or outside Blender.

A potential solution I can look into would be offloading the issue to a file: adding an `issue_file` parameter to the JSON. Then it could search for the file, either in the currently open file's directory or an absolute location, and run related functions from that Python file. Then you would have a normal Python file to do whatever at your heart's content. But then again, it's not that different from just making your own Python file and running that yourself. And you could save yourself the trouble of coupling it to BLint.

In short, I'll keep this open so people are aware of this known issue. And I'll look into the file solution above or take further implementation ideas. But the scope of BLint is limited, including the issues it expects to solve. Because most of us that _do_ code or need heavy scripting to resolve issues, we'd be better off running a custom script without having to deal with a dependency anyway 😄

**Comment by artificerchris** [comment]
Actually, I do just need to use it once per lint right now, so the `_import_` truck will work for me, and I totally get your reasoning on the rest of it and would like to say I'm in favor of the file approach, if you do add it, id much rather keep the lint code in an external file

**Comment by semagnum** [comment]
Sounds good, and I'll look into the file option.

**Comment by semagnum** [comment]
@artificerchris , I made a new branch [rule-file](https://github.com/semagnum/blint/tree/rule-file) to test this. It includes a new template/test file that you can use.

Feel free to review/test/debug it, see if it's what you'd expect :)

**Comment by artificerchris** [comment]
Firstly, thank you very much for implementing this!

I just went through and made a rule, and it looks like there's an Issue object that's expected to be returned, so I just copied the fields off your issue object and came up with this script, if you have an idea on how I could access your model from this standalone script, I think that would clean things up a lot (or just any way I could do this without needing to define the Issue class). The only other thing I'd add is a file picker button like the External Rules file has.

Here's the rule I made, it looks for a file named exactly like the blend file, but with `-preview.png` on the end instead of `.blend`:

```
import os
import bpy
from typing import List


class Issue:
    description: str
    severity_icon: str
    category_icon: str
    fix_expr: str
    rule_file: str

    def get(self, property: str, _: str = "") -> str:
        match property:
            case "description":
                return self.description
            case "severity_icon":
                return self.severity_icon
            case "category_icon":
                return self.category_icon
            case "fix_expr":
                return self.fix_expr
            case "rule_file":
                return self.rule_file
            case _:
                return ""


def get_issues() -> List[Issue]:
    issues = []
    filepath = bpy.data.filepath
    directory = os.path.dirname(filepath)
    basename = os.path.basename(filepath).replace(".blend", "-preview.png")
    preview_file = os.path.join(directory, "..", basename)
    if not os.path.exists(preview_file):
        issue = Issue()
        issue.description = f"Preview file '{preview_file}' does not exist."
        issue.severity_icon = "ERROR"
        issue.category_icon = "FILE"
        issue.fix_expr = ""
        issue.rule_file = __file__
        issues.append(issue)
    return issues
```

**Comment by semagnum** [comment]
The `external_rule_template.py` in the branch's commit has what you're looking for: https://github.com/semagnum/blint/blob/rule-file/external_rule_template.py

`get_issues()` just return a list of dicts, one for each issue. You don't need the issue class at all.

**Comment by semagnum** [comment]
Added the file subtype to the rule properties

**Comment by artificerchris** [comment]
Oops, I missed that file, now my external rules script looks like this, and seems to be working well: ```
import bpy
import os
from typing import List, Dict


def get_issues() -> List[Dict[str, str]]:
    issues: List[Dict[str, str]] = []

    result = rule_preview_file_exists()
    if result:
        issues.append(result)

    return issues


def fix_issues(*args):
    pass


def rule_preview_file_exists():
    filepath = bpy.data.filepath
    directory = os.path.dirname(filepath)
    basename = os.path.basename(filepath).replace(".blend", "-preview.png")
    preview_file = os.path.join(directory, "..", basename)

    if not os.path.exists(preview_file):
        return {
            "description": f"Preview file '{preview_file}' does not exist.",
            "severity_icon": "ERROR",
            "category_icon": "FILE",
            "fix_expr": "",
        }
    else:
        return None
```

The only other thing I noticed is that the file picker you added returns a path relative to the current blend file, instead of an absolute path, which I think will cause this rule to break if I open another file, so I think the external file path it picks should either be absolute or relative to the config json file.

Also, thank you again for the quick improvements and for taking my feedback, having these rule scripts will make blint much more useful for my team with our asset library.

**Comment by semagnum** [comment]
> The only other thing I noticed is that the file picker you added returns a path relative to the current blend file, instead of an absolute path, ... I think the external file path it picks should either be absolute or relative to the config json file.

Yeah, it needs to be absolute path. Unfortunately that's the Blender default when the property isn't saved under preferences, and I can't change that default directly afaik. However, you can workaround this by either directly pasting/typing the path directly into the field or, while in the file browser, opening the right side panel with the tiny arrow and disabling "Relative Path". 

![Blender's file browser: after clicking the little arrow on the right side, it will show a toggle for "Relative Path"](https://github.com/user-attachments/assets/1d70af85-1bc8-4653-87db-554cb8dd623b)

And fwiw you only need `fix_issues(*args)` if you plan to use it. The code checks if it exists before calling it; if it's not there, it'll default to the `fix_expr`s you pass for `get_issues()`.

**Comment by artificerchris** [comment]
Heh, just a habit from working in languages that freak if a symbol is missing, thanks for clarifying that! I'll be adding a bunch more rules over the next week or so, if I run into anything else, I'll provide more feedback / try to submit a PR.
